File: apps/mobile/lib/core/utils/acp_model_vendor.dart
import '../models/acp_models.dart';
import '../../l10n/generated/app_localizations.dart';

/// Vendors used to group Cursor ACP models instead of a flat list.
enum AcpModelVendor { anthropic, gpt, grok, google, other }

const kAcpModelVendors = [
  AcpModelVendor.anthropic,
  AcpModelVendor.gpt,
  AcpModelVendor.grok,
  AcpModelVendor.google,
  AcpModelVendor.other,
];

String _haystack(CursorModelOption model) =>
    '${model.id} ${model.displayName}'.toLowerCase();

/// Classify a Cursor ACP model into a vendor bucket (mirrors the desktop
/// classifier so the mobile picker keeps the provider → model hierarchy).
AcpModelVendor classifyAcpModelVendor(CursorModelOption model) {
  final text = _haystack(model);
  if (RegExp(r'\bgrok\b').hasMatch(text) || text.contains('xai')) {
    return AcpModelVendor.grok;
  }
  if (text.contains('gemini') ||
      text.contains('gemma') ||
      text.contains('google')) {
    return AcpModelVendor.google;
  }
  if (text.contains('claude') ||
      text.contains('anthropic') ||
      text.contains('sonnet') ||
      text.contains('opus') ||
      text.contains('haiku')) {
    return AcpModelVendor.anthropic;
  }
  if (text.contains('gpt') ||
      text.contains('openai') ||
      text.contains('chatgpt') ||
      RegExp(r'(^|[^a-z])o[1-9]([.-]|$)').hasMatch(text)) {
    return AcpModelVendor.gpt;
  }
  return AcpModelVendor.other;
}

Map<AcpModelVendor, List<CursorModelOption>> groupCursorModelsByVendor(
  List<CursorModelOption> models,
) {
  final grouped = {
    for (final vendor in kAcpModelVendors) vendor: <CursorModelOption>[],
  };
  for (final model in models) {
    grouped[classifyAcpModelVendor(model)]?.add(model);
  }
  return grouped;
}

String acpModelVendorLabel(AppLocalizations l10n, AcpModelVendor vendor) {
  switch (vendor) {
    case AcpModelVendor.anthropic:
      return 'Anthropic';
    case AcpModelVendor.gpt:
      return 'GPT';
    case AcpModelVendor.grok:
      return 'Grok';
    case AcpModelVendor.google:
      return 'Google';
    case AcpModelVendor.other:
      return l10n.modelCascadeVendorOther;
  }
}

File: apps/desktop/src/shared/acp-model-vendor.ts
export type AcpModelVendor = "anthropic" | "gpt" | "grok" | "google" | "other";

export interface AcpModelOption {
  id: string;
  displayName: string;
  current?: boolean;
  default?: boolean;
}

export const ACP_MODEL_VENDORS: readonly AcpModelVendor[] = ["anthropic", "gpt", "grok", "google", "other"];

export const ACP_MODEL_VENDOR_ICONS: Record<AcpModelVendor, string> = {
  anthropic: "./agent-icons/claude-code.ico",
  gpt: "./agent-icons/codex.ico",
  grok: "./agent-icons/grok.ico",
  google: "./agent-icons/gemini.png",
  other: "./agent-icons/other.svg",
};

function haystack(model: { id: string; displayName?: string }): string {
  return `${model.id} ${model.displayName ?? ""}`.toLowerCase();
}

export function classifyAcpModelVendor(model: { id: string; displayName?: string }): AcpModelVendor {
  const text = haystack(model);
  if (/\bgrok\b/.test(text) || text.includes("xai")) return "grok";
  if (text.includes("gemini") || text.includes("gemma") || text.includes("google")) return "google";
  if (
    text.includes("claude") ||
    text.includes("anthropic") ||
    text.includes("sonnet") ||
    text.includes("opus") ||
    text.includes("haiku")
  ) {
    return "anthropic";
  }
  if (
    text.includes("gpt") ||
    text.includes("openai") ||
    text.includes("chatgpt") ||
    /(^|[^a-z])o[1-9]([.-]|$)/.test(text)
  ) {
    return "gpt";
  }
  return "other";
}

export function groupAcpModelsByVendor<T extends { id: string; displayName?: string }>(
  models: readonly T[],
): Record<AcpModelVendor, T[]> {
  const grouped: Record<AcpModelVendor, T[]> = {
    anthropic: [],
    gpt: [],
    grok: [],
    google: [],
    other: [],
  };
  for (const model of models) {
    grouped[classifyAcpModelVendor(model)].push(model);
  }
  return grouped;
}

export function filterAcpModels<T extends { id: string; displayName?: string }>(
  models: readonly T[],
  query: string,
): T[] {
  const needle = query.trim().toLowerCase();
  if (!needle) return [...models];
  return models.filter((model) => haystack(model).includes(needle));
}

export function resolveAcpModelVendor(
  selectedModelId: string | undefined,
  models: readonly { id: string; displayName?: string }[],
): AcpModelVendor {
  if (selectedModelId) {
    const selected = models.find((model) => model.id === selectedModelId);
    return classifyAcpModelVendor(selected ?? { id: selectedModelId });
  }
  const grouped = groupAcpModelsByVendor(models);
  return ACP_MODEL_VENDORS.find((vendor) => grouped[vendor].length > 0) ?? "anthropic";
}

export function resolveAcpComposerTriggerLabel(
  models: readonly { id: string; displayName?: string; current?: boolean }[],
  selectedModelId: string | undefined,
  defaultLabel: string,
): string {
  const selectedId = selectedModelId?.trim();
  if (selectedId) {
    const match = models.find((model) => model.id === selectedId);
    const displayName = match?.displayName?.trim();
    return displayName || selectedId;
  }
  const current = models.find((model) => model.current);
  const currentName = current?.displayName?.trim();
  return currentName || defaultLabel;
}
